BasicHistorySessionsKotlin/app/src/main/java/com/google/android/gms/fit/samples/basichistorysessionskotlin/MainActivity.kt

```diff
@@ -0,0 +1,513 @@
+        initializeLogging()
+
+        // When permissions are revoked the app is restarted so onCreate is sufficient to check for
+        // permissions core to the Activity's functionality.
+
+        checkPermissionsAndRun(::insertAndVerifySession)
+    }
+
+    private fun checkPermissionsAndRun(signInAction: () -> Any) {
+        if (hasRuntimePermissions()) {
+            fitSignIn(signInAction)
+        } else {
+            requestRuntimePermissions(signInAction)
+        }
+    }
+
+    /**
+     * Checks that the user is signed in, and if so, executes the specified function. If the user is
+     * not signed in, initiates the sign in flow, specifying the post-sign in function to execute.
+     *
+     * @param signInAction The function to execute if already signed in, or once signed in.
+     */
+    private fun fitSignIn(signInAction: () -> Any) {
+        if (!hasOAuthPermissions()) {
+            val requestCode = signInActionMap[signInAction]
+            requestCode?.let {
+                GoogleSignIn.requestPermissions(
+                        this,
+                        requestCode,
+                        getGoogleAccount(), fitnessOptions)
+            }
+        } else signInAction()
+    }
+
+    /**
+     * Handles the callback from the OAuth sign in flow, executing the post sign in function
+     */
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        super.onActivityResult(requestCode, resultCode, data)
+
+        when (resultCode) {
+            RESULT_OK -> {
+                val postSignInAction = signInResponseMap[requestCode]
+                postSignInAction?.let {
+                    postSignInAction()
+                }
+            }
+            else -> oAuthErrorMsg(requestCode, resultCode)
+        }
+    }
+
+    private fun oAuthErrorMsg(requestCode: Int, resultCode: Int) {
+        val message = """
+            There was an error signing into Fit. Check the troubleshooting section of the README
+            for potential issues.
+            Request code was: $requestCode
+            Result code was: $resultCode
+        """.trimIndent()
+        Log.e(TAG, message)
+    }
+
+    private fun hasOAuthPermissions() = GoogleSignIn.hasPermissions(getGoogleAccount(), fitnessOptions)
+
+    private fun getGoogleAccount() = GoogleSignIn.getAccountForExtension(this, fitnessOptions)
+
+    private fun insertAndVerifySession() = insertSession().continueWith { verifySession() }
+
+    /**
+     * Creates and executes a [SessionInsertRequest] using [  ] to insert a session.
+     */
+    private fun insertSession(): Task<Void> {
+        //First, create a new session and an insertion request.
+        val insertRequest = createSessionInsertRequest()
+
+        // [START insert_session]
+        // Then, invoke the Sessions API to insert the session and await the result,
+        // which is possible here because of the AsyncTask. Always include a timeout when
```

Comment:
I'm confused by this. Where should you set the timeout then?

Reply:
This comment in the code is somewhat old I think - I should have spotted that. It doesn't bear reference to what is actually going on now... removed it.